pkg/actors/health/health.go

```diff
@@ -0,0 +1,297 @@
+	once    sync.Once
+	wg      sync.WaitGroup
+}
+
+func New(opts ...Option) (*Checker, error) {
+	options := &options{
+		failureThreshold:       defaultFailureThreshold,
+		initialDelay:           defaultInitialDelay,
+		requestTimeout:         defaultRequestTimeout,
+		successStatusCode:      defaultSuccessStatusCode,
+		healthyStateInterval:   defaultHealthyStateInterval,
+		unhealthyStateInterval: defaultUnHealthyStateInterval,
+		clock:                  new(clock.RealClock),
+	}
+
+	for _, o := range opts {
+		o(options)
+	}
+
+	if len(options.address) == 0 {
+		return nil, errors.New("required option 'address' is missing")
+	}
+
+	if options.client == nil {
+		options.client = new(http.Client)
+	}
+
+	return &Checker{
+		ch:                     make(chan bool),
+		client:                 options.client,
+		address:                options.address,
+		initialDelay:           options.initialDelay,
+		requestTimeout:         options.requestTimeout,
+		failureThreshold:       options.failureThreshold,
+		healthyStateInterval:   options.healthyStateInterval,
+		unhealthyStateInterval: options.unhealthyStateInterval,
+		successStatusCode:      options.successStatusCode,
+		clock:                  options.clock,
+		closeCh:                make(chan struct{}),
+	}, nil
+}
+
+func (c *Checker) Run(ctx context.Context) {
+	c.wg.Add(2)
+	defer c.Close()
+	defer c.wg.Done()
+
+	ctx, cancel := context.WithCancel(ctx)
+	go func() {
+		defer c.wg.Done()
+		select {
+		case <-ctx.Done():
+		case <-c.closeCh:
+		}
+		cancel()
+	}()
+
+	select {
+	case <-ctx.Done():
+		return
+	case <-c.clock.After(c.initialDelay):
+	}
+
+	for {
+		c.doUnHealthyStateCheck(ctx)
+		c.doHealthyStateCheck(ctx)
+
+		select {
+		case <-ctx.Done():
+			return
+		default:
+		}
+	}
+}
+
+func (c *Checker) HealthChannel() <-chan bool {
+	return c.ch
+}
+
+func (c *Checker) Close() {
+	defer c.wg.Wait()
+	c.once.Do(func() {
+		close(c.closeCh)
+		c.client.CloseIdleConnections()
+		// Wait for health reports to finish before closing the channel.
+		c.wg.Wait()
+		close(c.ch)
+	})
+}
+
+func (c *Checker) doUnHealthyStateCheck(ctx context.Context) {
+	if ctx.Err() != nil {
+		return
+	}
+
+	ticker := c.clock.NewTicker(c.unhealthyStateInterval)
+	defer ticker.Stop()
+
+	for {
+		// Immediately get health status if in unhealthy state.
+		if c.getStateHealth(ctx) {
+			healthLogger.Info("Actor health check succeeded, marking healthy")
+			c.reportHealth(ctx, true)
+			return
+		}
+
+		select {
+		case <-ctx.Done():
+			return
+
+		case <-ticker.C():
+		}
+	}
+}
+
+func (c *Checker) doHealthyStateCheck(ctx context.Context) {
+	if ctx.Err() != nil {
+		return
+	}
+
+	ticker := c.clock.NewTicker(c.healthyStateInterval)
+	defer ticker.Stop()
+
+	var failureCount int32
+	for {
+		select {
+		case <-ctx.Done():
+			return
+
+		case <-ticker.C():
+			if c.getStateHealth(ctx) {
+				failureCount = 0
+			} else {
+				failureCount++
+			}
+
+			if failureCount >= c.failureThreshold {
```

Comment:
This is to prevent overflows

```suggestion
			if failureCount >= c.failureThreshold {
                failureCount = c.failureThreshold
```

Reply:
Why would this prevent an overflow?

Reply:
Sorry, still not quite sure how this would be problematic. `failureCount` is a variable local to the function which we use to track how many failures were encountered doing the health check. This begins at 0 and once it hits the threshold, we report the unhealthy state and return from the function to enter the unhealthy state. This throws away the variable, where it will be 0 again when we re-enter.

It is actually impossible for `failureCount` to be any larger than `failureThreshold` (we could change `>=` to just `==`), and so it's impossible for it to overflow, since  `failureThreshold` _has_ to be at most the maximum value that `failureCount` could be.

Reply:
Ok, I misunderstood this code